### PR TITLE
Xiangyu/shared variable

### PR DESCRIPTION
**Pull request by Xiangyu-Hu**
1. Using unique pointer for shared particle data.
2. Simplify the type for contact figuration.
3. Renaming resize configuration and unique pointers keeper for better expression.

### CODE REVIEW

**Comment by FabienPean-Virtonomy** [comment]
Renaming generates a lot of noise in the PR and drowns the more meaningful changes, please consider splitting those type of changes for future PRs.

File: SPHINXsys/src/shared/particles/base_particles.h [hunk 1]
```diff
@@ -82,7 +82,7 @@ namespace SPH
 	class BaseParticles
 	{
 	private:
-		UniquePtrKeepers<BaseDynamics<void>> derived_particle_data_; /**< Unique ptr for Base dynamics. */
+		UniquePtrsKeeper<BaseDynamics<void>> derived_particle_data_; /**< Unique ptr for Base dynamics. */
 		
 		template <typename VariableType>
 		struct AllVariablesDelete
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/shared/particles/base_particles.h:88]: `AllVariablesDelete` should now be removed according to current changes

File: SPHINXsys/src/shared/body_relations/base_body_relation.h [hunk 2]
```diff
@@ -143,18 +143,19 @@ namespace SPH
 	class BaseContactRelation : public SPHRelation
 	{
 	protected:
+		UniquePtrsKeeper<ParticleConfiguration> configuration_ptrs_keeper_;
 		virtual void resetNeighborhoodCurrentSize();
 
 	public:
 		RealBodyVector contact_bodies_;
-		ContactParticleConfiguration contact_configuration_; /**< Configurations for particle interaction between bodies. */
+		StdVec<ParticleConfiguration *> contact_configuration_; /**< Configurations for particle interaction between bodies. */
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/shared/body_relations/base_body_relation.h:151]: There is no visible benefits to this. Value semantic is preferable, it just needs a few methods around the original container to enforce security and proper usage.

**Review by FabienPean-Virtonomy — CHANGES_REQUESTED** [review]

File: SPHINXsys/src/shared/particles/base_particles.h (hunk 1 above)
Comment (Xiangyu-Hu) [reply SPHINXsys/src/shared/particles/base_particles.h:88]: Ok.

File: SPHINXsys/src/shared/body_relations/base_body_relation.h (hunk 2 above)
Comment (FabienPean-Virtonomy) [reply SPHINXsys/src/shared/body_relations/base_body_relation.h]: Let me reformulate, what I meant is that I don't see the reason to replace `ContactParticleConfiguration` by `UniquePtrsKeeper<ParticleConfiguration>`, while making the member variable protected/private with proper set of getter would do the same.

File: SPHINXsys/src/shared/body_relations/base_body_relation.h [hunk 3]
```diff
@@ -143,6 +143,7 @@ namespace SPH
 	class BaseContactRelation : public SPHRelation
 	{
 	protected:
+		UniquePtrsKeeper<ParticleConfiguration> configuration_ptrs_keeper_;
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/shared/body_relations/base_body_relation.h:146]: This

**Review by FabienPean-Virtonomy — APPROVED** [review]

**Pull request merged**